File: examples/rust-map/src/render_target.rs
#[cfg(maplibre_render_backend = "metal")]
mod metal_target;
#[cfg(maplibre_render_backend = "opengl")]
mod opengl_target;
#[cfg(maplibre_render_backend = "vulkan")]
mod vulkan_target;

#[cfg(maplibre_render_backend = "metal")]
pub use metal_target::RenderTarget;
#[cfg(maplibre_render_backend = "opengl")]
pub use opengl_target::RenderTarget;
#[cfg(maplibre_render_backend = "vulkan")]
pub use vulkan_target::RenderTarget;

use std::error::Error as StdError;

use maplibre_native_ffi::{
    AcquiredFrameHandle, Error, ErrorKind, FrameDemand, FrameDisposition, FrameGpuSync,
    NativeFuture, RenderSessionAttachment, RenderSessionHandle, RenderTargetExtent,
};

use crate::viewport::Viewport;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Mode {
    OwnedTexture,
    BorrowedTexture,
    NativeSurface,
}

impl Mode {
    pub fn cli_name(self) -> &'static str {
        match self {
            Self::OwnedTexture => "owned-texture",
            Self::BorrowedTexture => "borrowed-texture",
            Self::NativeSurface => "native-surface",
        }
    }

    pub fn status(self) -> &'static str {
        match self {
            Self::OwnedTexture => "samples MapLibre-owned texture frames into the host swapchain",
            Self::BorrowedTexture => {
                "renders into a host-owned texture, then samples it into the host swapchain"
            }
            Self::NativeSurface => "renders directly to the host window surface",
        }
    }

    pub fn parse(value: &str) -> Result<Self, String> {
        match value {
            "owned-texture" => Ok(Self::OwnedTexture),
            "borrowed-texture" => Ok(Self::BorrowedTexture),
            "native-surface" => Ok(Self::NativeSurface),
            _ => Err(format!("unknown render target '{value}'")),
        }
    }
}

/// One frame demand's outcome: whether the session rendered the demand, and
/// whether the map asked for another frame while it rendered this one.
#[derive(Clone, Copy, Debug, Default)]
pub struct FrameOutcome {
    pub rendered: bool,
    pub needs_repaint: bool,
}

pub fn extent(viewport: Viewport) -> RenderTargetExtent {
    RenderTargetExtent::new(
        viewport.logical_width,
        viewport.logical_height,
        viewport.scale_factor,
    )
}

/// A caller-graphics-thread render session plus the monotonic demand tokens
/// that tie each frame result back to the demand that produced it.
///
/// Every ordered session submission runs on the graphics thread that attached,
/// so this type also owns the loop that services driver work while one is
/// outstanding.
pub struct FrameDriver {
    session: RenderSessionHandle,
    next_token: u64,
}

impl FrameDriver {
    /// Services driver work until the attachment resolves.
    pub fn new(attachment: RenderSessionAttachment) -> maplibre_native_ffi::Result<Self> {
        let driver = Self {
            session: attachment.session,
            next_token: 0,
        };
        driver.drive(&attachment.completion)?;
        Ok(driver)
    }

    pub fn session(&self) -> &RenderSessionHandle {
        &self.session
    }

    /// Services driver work until one ordered submission resolves.
    pub fn drive(&self, operation: &NativeFuture<()>) -> maplibre_native_ffi::Result<()> {
        while !operation.is_ready() {
            if self.session.service_driver_work(0)? == 0 {
                std::thread::yield_now();
            }
        }
        operation.take()
    }

    /// Carries a new logical extent to the map through the attached session,
    /// which is the only extent authority while it stays attached.
    pub fn resize(&self, viewport: Viewport) -> maplibre_native_ffi::Result<()> {
        let operation = self.session.resize(&extent(viewport))?;
        self.drive(&operation)
    }

    /// Submits one demand, services driver work, and reports the outcome of
    /// the result carrying this demand's token.
    pub fn render_frame(&mut self, present: bool) -> maplibre_native_ffi::Result<FrameOutcome> {
        self.next_token += 1;
        let token = self.next_token;
        self.session.request_frame(FrameDemand {
            present,
            token,
            ..FrameDemand::default()
        })?;
        self.session.service_driver_work(0)?;
        Ok(self
            .session
            .drain_frame_results()?
            .into_iter()
            .find(|result| result.token == token)
            .map(|result| FrameOutcome {
                rendered: result.disposition == FrameDisposition::Rendered,
                needs_repaint: result.needs_repaint,
            })
            .unwrap_or_default())
    }

    /// Leases the rendered frame, reporting `None` while the ring holds none.
    pub fn acquire_frame(&self) -> maplibre_native_ffi::Result<Option<AcquiredFrameHandle>> {
        match self.session.acquire_frame() {
            Ok(frame) => Ok(Some(frame)),
            Err(error) if error.kind() == ErrorKind::NotReady => Ok(None),
            Err(error) => Err(error),
        }
    }

    /// Detaches on the graphics thread, then destroys the session.
    pub fn close(self) -> Result<(), Box<dyn StdError>> {
        let operation = self.session.detach()?;
        self.drive(&operation)?;
        self.session
            .destroy()
            .map_err(|error| Box::new(error) as Box<dyn StdError>)
    }
}

pub fn require_cpu_complete_producer(
    frame: &AcquiredFrameHandle,
) -> maplibre_native_ffi::Result<()> {
    let producer = frame.producer_sync()?;
    if matches!(producer, FrameGpuSync::CpuComplete) {
        return Ok(());
    }
    Err(Error::new(
        ErrorKind::InvalidState,
        None,
        format!("rust-map cannot consume a {producer:?} producer synchronization payload"),
    ))
}

pub fn compositor_error(message: impl Into<String>) -> Error {
    Error::new(ErrorKind::NativeError, None, message)
}

File: examples/zig-map/render_target.zig
const std = @import("std");

const maplibre = @import("maplibre_native_ffi");
const diagnostics = @import("diagnostics.zig");
const types = @import("types.zig");

/// One frame demand's outcome: whether the session rendered the demand, and
/// whether the map asked for another frame while it rendered this one.
pub const FrameOutcome = struct {
    rendered: bool = false,
    needs_repaint: bool = false,
};

/// An attached render session plus the monotonic demand tokens that tie each
/// frame result back to the demand that produced it.
pub const Session = struct {
    pub const Target = union(enum) {
        none,
        texture: maplibre.RenderSessionHandle,
        surface: maplibre.RenderSessionHandle,
    };

    /// The one ordered submission that can be outstanding: attach, resize,
    /// target replacement, or detach.
    const Pending = struct {
        future: maplibre.Future(void),
        app_error: types.AppError,
        message: []const u8,
    };

    target: Target = .none,
    /// The map this session renders. Target replacement changes only the
    /// graphics resource, so those paths carry the extent to the map directly.
    map: ?*maplibre.MapHandle = null,
    next_token: u64 = 0,
    pending: ?Pending = null,

    pub fn deinit(self: *Session) void {
        const handle = self.nativeHandle() orelse {
            self.discardPending();
            return;
        };
        // The outstanding submission owns the completion slot the detach
        // needs, so it finishes first.
        var abandon = if (self.awaitPending()) |_| false else |_| true;
        if (!abandon) {
            if (handle.detach()) |completion| {
                self.beginPending(completion, self.detachError(), "render session detach failed");
                abandon = if (self.awaitPending()) |_| false else |_| true;
            } else |_| {
                abandon = true;
            }
        }
        if (abandon) _ = handle.abandon() catch {};
        handle.destroy() catch {};
        self.discardPending();
        self.target = .none;
    }

    pub fn nativeHandle(self: *Session) ?*maplibre.RenderSessionHandle {
        return switch (self.target) {
            .none => null,
            .texture => |*value| value,
            .surface => |*value| value,
        };
    }

    pub fn textureHandle(self: *Session) !*maplibre.RenderSessionHandle {
        return switch (self.target) {
            .texture => |*texture| texture,
            .none, .surface => types.AppError.TextureResizeFailed,
        };
    }

    pub fn surfaceHandle(self: *Session) !*maplibre.RenderSessionHandle {
        return switch (self.target) {
            .surface => |*surface| surface,
            .none, .texture => types.AppError.SurfaceAttachFailed,
        };
    }

    /// Takes ownership of the future one ordered submission returned, so the
    /// render loop can drive it instead of blocking the caller.
    pub fn beginPending(
        self: *Session,
        future: maplibre.Future(void),
        app_error: types.AppError,
        message: []const u8,
    ) void {
        std.debug.assert(self.pending == null);
        self.pending = .{ .future = future, .app_error = app_error, .message = message };
    }

    /// Services caller-driver work and reports whether the outstanding
    /// submission is still pending.
    pub fn poll(self: *Session) !bool {
        if (self.pending == null) return false;
        const session = self.nativeHandle() orelse {
            self.discardPending();
            return false;
        };
        var serviced = true;
        _ = session.serviceDriverWork(0) catch |err| {
            serviced = false;
            diagnostics.logError(self.pending.?.message, err, null);
        };
        if (serviced and !try self.pending.?.future.poll()) return true;
        var pending = self.takePending().?;
        defer pending.future.deinit();
        if (!serviced) return pending.app_error;
        pending.future.wait(null) catch |err| {
            diagnostics.logError(pending.message, err, null);
            return pending.app_error;
        };
        return false;
    }

    /// Services caller-driver work until the outstanding submission completes.
    /// Startup and shutdown block here; the render loop polls instead.
    pub fn awaitPending(self: *Session) !void {
        while (try self.poll()) {}
    }

    /// Starts the session resize that carries the new logical extent to the
    /// map. The render loop drives it to completion through `poll`.
    pub fn startResize(self: *Session, viewport: types.Viewport) !void {
        const session = self.nativeHandle() orelse return types.AppError.TextureResizeFailed;
        const app_error = self.resizeError();
        const completion = session.resize(extent(viewport)) catch |err| {
            diagnostics.logError("render target resize failed", err, null);
            return app_error;
        };
        self.beginPending(completion, app_error, "render target resize failed");
    }

    /// Carries the new logical extent to the map on the paths where the
    /// session cannot: a caller-owned texture the host sizes, and a replaced
    /// surface target. Both change only the graphics resource.
    pub fn resizeMap(self: *Session, viewport: types.Viewport) !void {
        const map = self.map orelse return self.resizeError();
        var completion = map.resize(
            viewport.logical_width,
            viewport.logical_height,
            viewport.scale_factor,
        ) catch |err| {
            diagnostics.logError("map resize failed", err, null);
            return self.resizeError();
        };
        completion.deinit();
    }

    /// Submits one frame demand, services caller-driver work, and reports the
    /// outcome of the result carrying this demand's token.
    pub fn renderUpdate(
        self: *Session,
        allocator: std.mem.Allocator,
        diagnostic_store: ?*const maplibre.DiagnosticStore,
    ) !FrameOutcome {
        const presents = self.target == .surface;
        const session = self.nativeHandle() orelse return FrameOutcome{};
        const app_error = if (presents)
            types.AppError.SurfaceRenderFailed
        else
            types.AppError.TextureRenderFailed;
        self.next_token += 1;
        const token = self.next_token;
        session.requestFrame(.{ .if_needed = true, .present = presents, .token = token }) catch |err| {
            diagnostics.logError("frame request failed", err, diagnostic_store);
            return app_error;
        };
        _ = session.serviceDriverWork(0) catch |err| {
            diagnostics.logError("render driver service failed", err, diagnostic_store);
            return app_error;
        };
        // The drain copies its results, so this loop owns them for the frame.
        var batch = try session.drainFrameResults(allocator);
        defer batch.deinit();
        for (0..batch.len()) |index| {
            const result = try batch.at(index);
            if (result.token != token) continue;
            return .{
                .rendered = result.disposition == .rendered,
                .needs_repaint = result.needs_repaint,
            };
        }
        return .{};
    }

    fn resizeError(self: *const Session) types.AppError {
        return if (self.target == .surface)
            types.AppError.SurfaceResizeFailed
        else
            types.AppError.TextureResizeFailed;
    }

    fn detachError(self: *const Session) types.AppError {
        return if (self.target == .surface)
            types.AppError.SurfaceAttachFailed
        else
            types.AppError.TextureAttachFailed;
    }

    fn takePending(self: *Session) ?Pending {
        const pending = self.pending;
        self.pending = null;
        return pending;
    }

    fn discardPending(self: *Session) void {
        if (self.takePending()) |pending| {
            var owned = pending;
            owned.future.deinit();
        }
    }
};

pub fn textureSession(map: *maplibre.MapHandle, attachment: maplibre.RenderSessionAttachment) !Session {
    return attachedSession(map, attachment, .{ .texture = attachment.session });
}

pub fn surfaceSession(map: *maplibre.MapHandle, attachment: maplibre.RenderSessionAttachment) !Session {
    return attachedSession(map, attachment, .{ .surface = attachment.session });
}

/// Services caller-driver work until the attachment resolves, abandoning the
/// session when it does not.
fn attachedSession(
    map: *maplibre.MapHandle,
    attachment: maplibre.RenderSessionAttachment,
    target: Session.Target,
) !Session {
    var owned = attachment;
    var session = Session{ .target = target, .map = map };
    errdefer {
        _ = owned.session.abandon() catch {};
        owned.session.destroy() catch {};
    }
    session.beginPending(owned.attached, session.detachError(), "render target attach failed");
    try session.awaitPending();
    return session;
}

pub fn extent(viewport: types.Viewport) maplibre.RenderTargetExtent {
    return .{ .width = viewport.logical_width, .height = viewport.logical_height, .scale_factor = viewport.scale_factor };
}
